engine/execution/ingestion/core.go

```diff
@@ -0,0 +1,461 @@
+			Int("missing_col", len(missingColls)).
+			Int("executables", len(executables)).
+			Msgf("block is enqueued")
+
+		return missingColls, executables, nil
+	}
+
+	// handle exception
+	if !errors.Is(err, storage.ErrNotFound) {
+		return nil, nil, fmt.Errorf("failed to get parent state commitment for block %v: %w", block.Header.ParentID, err)
+	}
+
+	// the parent block is an unexecuted block.
+	// we can enqueue the block without providing the state commitment
+	missingColls, executables, err := e.blockQueue.HandleBlock(block, nil)
+	if err != nil {
+		if !errors.Is(err, block_queue.ErrMissingParent) {
+			return nil, nil, fmt.Errorf("unexpected error while adding block to block queue: %w", err)
+		}
+
+		// if parent is missing, there are two possibilities:
+		// 1) parent was never enqueued to block queue
+		// 2) parent was enqueued, but it has been executed and removed from the block queue
+		// however, actually 1) is not possible 2) is the only possible case here, why?
+		// because forwardProcessableToHandler guarantees we always enqueue a block before its child,
+		// which means when HandleBlock is called with a block, then its parent block must have been
+		// called with HandleBlock already. Therefore, 1) is not possible.
+		// And the reason 2) is possible is because the fact that its parent block is missing
+		// might be outdated since OnBlockExecuted might be called concurrently in a different thread.
+		// it means OnBlockExecuted is called in a different thread after us getting the parent commit
+		// and before HandleBlock was called, therefore, we should re-enqueue the block with the
+		// parent commit. It's necessary to check again whether the parent block is executed after the call.
+		lg.Warn().Msgf(
+			"block is missing parent block, re-enqueueing %v(parent: %v)",
+			blockID, block.Header.ParentID,
+		)
+
+		parentCommitment, err := e.execState.StateCommitmentByBlockID(block.Header.ParentID)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get parent state commitment when re-enqueue block %v (parent:%v): %w",
+				blockID, block.Header.ParentID, err)
+		}
+
+		// now re-enqueue the block with parent commit
+		missing, execs, err := e.blockQueue.HandleBlock(block, &parentCommitment)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unexpected error while reenqueue block to block queue: %w", err)
+		}
+
+		missingColls = deduplicate(append(missingColls, missing...))
+		executables = deduplicate(append(executables, execs...))
+	}
+
+	lg.Info().Bool("parent_is_executed", false).
+		Int("missing_col", len(missingColls)).
+		Int("executables", len(executables)).
+		Msgf("block is enqueued")
+
+	return missingColls, executables, nil
+}
+
+func (e *Core) onBlockExecuted(
+	block *entity.ExecutableBlock,
+	computationResult *execution.ComputationResult,
+	startedAt time.Time,
+) error {
+	commit := computationResult.CurrentEndState()
+
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	defer wg.Wait()
```

Comment:
should this be synchronized so `BeforeComputationResultSaved` is guaranteed to be called before `OnComputationResultSaved`? it's currently possible for them to be out of order

Reply:
This does not need to be synchronized, in fact we want them to be performed parallel. 

The use case is to performance some action while saving the execution result. In our scenario, it's to upload the execution result to GCP (which is deprecated). Since the execution result is deterministic, it's OK to upload while saving it to database. I called it `Before...Saved` just to indicate this callback can be used to performance actions that are parallel to saving the execution result.